### DISCUSSION TITLE
webpack is not noticing changes to lib delibered via commit id

### DISCUSSION THREAD

**Discussion by opensourcerebel** [webpack/webpack#16818] [Q&A]
# Bug report

**What is the current behavior?**
1. Initial state
I am using angular + yarn + a dependecy which is delievered via git commit e.g.
```json
"dependencies": {
    "@myorg/myproj": "git+https://myproj.git#somehash1111",
}
```
then I do a `yarn install` and `ng serve` and all is good. 
2. Make a change in the `myproj` dependency and commit it inside the `myproj` repo. After the commit a build is triggered for this dependency and the `dist` folder is uploaded as a commit in the `https://myproj.git` project (the `dist` folder has essentialy the output of the `ng build` command).
3. Consume the new version `somehash2222` by changing the `package.json` to
```json
"dependencies": {
    "@myorg/myproj": "git+https://someurl.git#somehash222",
}
```
3. Run `yarn install` and `ng serve`. The content of the old `somehas1111` is delivered. To resolve the issue one needs to clean the angular cache (which in turn cleans the `webpack` cache from the `.angular` folder).

Now at that point I know that `node_modules` is considered immutable and to my understanding the new `immutable` version of the `@myorg/myproj` is correctly delivered via a version change `somehash1111` -> `somehash2222`. To permanently workaround that I have used the [managedpaths](https://webpack.js.org/configuration/other-options/#managedpaths) option to instruct `webpack` to start observing this dir via a regexp similar to the `@asure` example. Which it does and now whenever someone changes the hash delievered dependency it is automatically picked up. But according to my understanding the additional config shall not be needed as there is a clear version change. Can it be that 'webpack' is confused as this is not a semver or because it does not look as dir?

**If the current behavior is a bug, please provide the steps to reproduce.**
See steps 1 to 3 for retproduction

**What is the expected behavior?**
when I change the dependency from ` "@myorg/myproj": "git+https://myproj.git#somehash1111"` to ` "@myorg/myproj": "git+https://myproj.git#somehash2222"` would be nice for webpack to notice the change then rebuild and most importantly update the cache.
 

**Other relevant information:**
webpack version:  (should be 5.75.0)
We are using it from angular 15 like that (snippet from yarn.lock)
```json
"@angular-devkit/build-angular@npm:15.2.1":
  version: 15.2.1
  resolution: "@angular-devkit/build-angular@npm:15.2.1"
  dependencies:
    "@ampproject/remapping": 2.2.0
    "@angular-devkit/architect": 0.1502.1
    "@angular-devkit/build-webpack": 0.1502.1
    "@angular-devkit/core": 15.2.1
    "@babel/core": 7.20.12
    "@babel/generator": 7.20.14
    "@babel/helper-annotate-as-pure": 7.18.6
    "@babel/helper-split-export-declaration": 7.18.6
    "@babel/plugin-proposal-async-generator-functions": 7.20.7
    "@babel/plugin-transform-async-to-generator": 7.20.7
    "@babel/plugin-transform-runtime": 7.19.6
    "@babel/preset-env": 7.20.2
    "@babel/runtime": 7.20.13
    "@babel/template": 7.20.7
    "@discoveryjs/json-ext": 0.5.7
    "@ngtools/webpack": 15.2.1
    ansi-colors: 4.1.3
    autoprefixer: 10.4.13
    babel-loader: 9.1.2
    babel-plugin-istanbul: 6.1.1
    browserslist: 4.21.5
    cacache: 17.0.4
    chokidar: 3.5.3
    copy-webpack-plugin: 11.0.0
    critters: 0.0.16
    css-loader: 6.7.3
    esbuild: 0.17.8
    esbuild-wasm: 0.17.8
    glob: 8.1.0
    https-proxy-agent: 5.0.1
    inquirer: 8.2.4
    jsonc-parser: 3.2.0
    karma-source-map-support: 1.4.0
    less: 4.1.3
    less-loader: 11.1.0
    license-webpack-plugin: 4.0.2
    loader-utils: 3.2.1
    magic-string: 0.29.0
    mini-css-extract-plugin: 2.7.2
    open: 8.4.1
    ora: 5.4.1
    parse5-html-rewriting-stream: 7.0.0
    piscina: 3.2.0
    postcss: 8.4.21
    postcss-loader: 7.0.2
    resolve-url-loader: 5.0.0
    rxjs: 6.6.7
    sass: 1.58.1
    sass-loader: 13.2.0
    semver: 7.3.8
    source-map-loader: 4.0.1
    source-map-support: 0.5.21
    terser: 5.16.3
    text-table: 0.2.0
    tree-kill: 1.2.2
    tslib: 2.5.0
    webpack: 5.75.0
    webpack-dev-middleware: 6.0.1
    webpack-dev-server: 4.11.1
    webpack-merge: 5.8.0
    webpack-subresource-integrity: 5.1.0
```
Node.js version: v19.6.1
Operating System: Ubuntu 20.04
Additional tools: yarn 3.4.1

**Comment by alexander-akait** [comment]
Can you provide your configuration, any change in package.json will invalidate cache (if it was not disabled)

**Comment by opensourcerebel** [comment]
@alexander-akait it will take me awhile to build and minimalistic example that is representative of my setup. 

In the meantime is there something I can do to debug webpack on my machine (sort of inplace debugging as the issue is 100% reproducible on my side) . I have tried adding 
```json
  infrastructureLogging: {
    appendOnly: true,
    level: 'verbose',
    debug:  /PackFileCache/
  },
```
to my custom webpack config but it did not said anything useful. I want to have some debugging output from the place where the decision is taken should the version be considered changed or not (on my side I have both `package.json` and `yarn.lock`). Or if you point me to some file where I can add manually additional logging?

**Reply by alexander-akait** [reply]
> o my custom webpack config but it did not said anything useful.

What output do you have?

Can you show me options of https://webpack.js.org/configuration/cache/#cache?

**Reply by opensourcerebel** [reply]
Before the issue there was no custom `webpack` as I were replying on angular settings like that 
```json
"cli": {
    "analytics": false,
    "cache": {
      "enabled": true,
      "environment": "all"
    }
  }
```
(I guess they are passing something to webpack a quick search in the `angular-cli` repo revelated this: https://github.com/angular/angular-cli/tree/15.1.x/packages/angular_devkit/build_angular/src/webpack/configs). After my change I had to tell angular that I would like to use a custom webpack config in `angular.js` (the architect->build->options). As of today the file looks like that:
```json
module.exports = {
  snapshot: {
    managedPaths: [
      /^(.+?[\\/]node_modules[\\/](?!(@myorg[\\/]myproj))(@.+?[\\/])?.+?)[\\/]/,
      /^(.+?[\\/]node_modules[\\/](?!(@myorg[\\/]myproj1))(@.+?[\\/])?.+?)[\\/]/,
      /^(.+?[\\/]node_modules[\\/](?!(@myorg[\\/]myproj2))(@.+?[\\/])?.+?)[\\/]/,
    ],
  },
};
```

**Reply by opensourcerebel** [reply]
One more thing: The lib project delivers its own `package.json` file but inside the `version` attribute does not change. f.e. the commits change but the `"version": "0.0.1",` remains. Can this be the issue? I tried searching inside the `webpack` code to find out the place where the decision is taken if a lib from `node_modules` is changed but as of now I could not pinpoint it.

**Reply by opensourcerebel** [reply]
aha ... can this be the problem? 
https://github.com/webpack/webpack/blob/v5.75.0/lib/FileSystemInfo.js#L3559

**Reply by alexander-akait** [reply]
Hm, might, so you don't change in package.json after do some changes in your library?

**Reply by opensourcerebel** [reply]
mhm yeah ... historical reasons (lots of politics ...)

**Reply by alexander-akait** [reply]
In this case, it can be problem, because webpack is thinking a package has the same content, there are some solutions:
1. https://webpack.js.org/configuration/cache/#cacheversion you can hash your `package.json` and use it in version, so any change will be invalid cache (not effective, but if you don't have it very often there is no problems)
2. Exclude your package from https://webpack.js.org/configuration/cache/#cachemanagedpaths, somewhere I provide examples good reexps here, but I think it will not a problem to write them
3. You can add https://webpack.js.org/configuration/cache/#cachebuilddependencies your package there, should work but I don't tests, maybe you still need to exclude it from https://webpack.js.org/configuration/cache/#cachemanagedpaths
